// @tact-lang/compiler/dist/node.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createNodeFileSystem = exports.run = void 0;
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
const parseConfig_1 = require("./config/parseConfig");
const createNodeFileSystem_1 = require("./vfs/createNodeFileSystem");
const build_1 = require("./pipeline/build");
const logger_1 = require("./logger");
async function configForSingleFile(fileName) {
    return {
        projects: [
            {
                name: path_1.default.basename(fileName, ".tact"),
                path: fileName,
                output: process.cwd(),
                options: { debug: true },
                mode: "full",
            },
        ],
        rootPath: process.cwd(),
        singleFile: true,
    };
}
async function loadConfig(fileName, configPath) {
    if (fileName)
        return configForSingleFile(fileName);
    if (!configPath)
        return null;
    let config;
    // Load config
    const resolvedPath = path_1.default.resolve(configPath);
    const rootPath = path_1.default.dirname(resolvedPath);
    if (!fs_1.default.existsSync(resolvedPath)) {
        console.warn("Unable to find config file at " + resolvedPath);
        return null;
    }
    try {
        config = (0, parseConfig_1.parseConfig)(fs_1.default.readFileSync(resolvedPath, "utf8"));
    }
    catch (e) {
        console.log(e);
        console.warn("Unable to parse config file at " + resolvedPath);
        return null;
    }
    return { singleFile: false, rootPath, ...config };
}
async function run(args) {
    const configWithRootPath = await loadConfig(args.fileName, args.configPath);
    if (!configWithRootPath) {
        return false;
    }
    // Resolve projects
    let projects = configWithRootPath.projects;
    if (args.projectNames && args.projectNames.length > 0) {
        // Check that all project names are valid
        for (const pp of args.projectNames) {
            if (!projects.find((v) => v.name === pp)) {
                console.warn("Unable to find project " + pp);
                return false;
            }
        }
        // Filter by names
        projects = projects.filter((v) => args.projectNames.includes(v.name));
    }
    if (projects.length === 0) {
        console.warn("No projects to compile");
        return false;
    }
    // Compile
    let success = true;
    const project = (0, createNodeFileSystem_1.createNodeFileSystem)(configWithRootPath.rootPath, false);
    const stdlib = (0, createNodeFileSystem_1.createNodeFileSystem)(path_1.default.resolve(__dirname, "..", "stdlib"), false); // Improves developer experience
    for (const config of projects) {
        logger_1.consoleLogger.log("💼 Compiling project " + config.name + "...");
        let cliConfig = { ...config };
        if (args.additionalCliOptions !== undefined &&
            args.additionalCliOptions.mode !== undefined) {
            cliConfig = { ...config, ...args.additionalCliOptions };
        }
        const built = await (0, build_1.build)({
            config: cliConfig,
            project,
            stdlib,
            logger: logger_1.consoleLogger,
        });
        success = success && built;
    }
    return success;
}
exports.run = run;
var createNodeFileSystem_2 = require("./vfs/createNodeFileSystem");
Object.defineProperty(exports, "createNodeFileSystem", { enumerable: true, get: function () { return createNodeFileSystem_2.createNodeFileSystem; } });
